File: task-sql/src/main/scala/com/zxkw/flume/interceptor/userprofile/app/TaskSQL.scala
package com.zxkw.flume.interceptor.userprofile.app

import java.util.Properties
import com.zxkw.flume.interceptor.userprofile.common.bean.{TagInfo, TaskInfo, TaskTagRule}
import com.zxkw.flume.interceptor.userprofile.common.constants.ConstCode
import com.zxkw.flume.interceptor.userprofile.common.dao.{TagInfoDAO, TaskInfoDAO, TaskTagRuleDAO}
import com.zxkw.flume.interceptor.userprofile.common.util.{MyPropertiesUtil, MySqlUtil}
import org.apache.spark.SparkConf
import org.apache.spark.sql.SparkSession


////  负责执行所有SQL类型的标签任务
object TaskSQL {


  //  1   获得标签的定义、规则、SQL 、四级标签的映射关系
  //  从mysql中取
  //
  //  2    从数仓中读取数据
  //
  //  3   写入到某个标签表
  //    3.1 定义标签表
  //    3.2 写入标签表
  def main(args: Array[String]): Unit = {
    //  1   获得标签的定义、规则、SQL 、四级标签的映射关系
    //  从my想·sql中取
    //  1.1  如何查询 mysql  工具类
    //  1.2 查询对应的表  tag_info 、task_info、task_tag_rule
    val sparkConf: SparkConf = new SparkConf().setAppName("task_sql_app")//.setMaster("local[*]")
    val sparkSession: SparkSession = SparkSession.builder().config(sparkConf).enableHiveSupport().getOrCreate()

    val taskId: String = args(0)
    val taskDate: String = args(1)

    // tag_info
    val tagInfo: TagInfo = TagInfoDAO.getTagInfoByTaskId(taskId)
    println(tagInfo)
    //task_info
    val taskInfo: TaskInfo = TaskInfoDAO.getTaskInfo(taskId)
    println(taskInfo)
    //task_tag_rule
    val taskTagRuleList: List[TaskTagRule] = TaskTagRuleDAO.getTaskTagRuleListByTaskId(taskId)
    println(taskTagRuleList)

    //  2定义目标表  create table   ( 1  一次性建立  （只要表结构不变）   2 周期性执行     a手工 b 程序)
    //   hive
    val tableName = tagInfo.tagCode.toLowerCase()
    val tagValueType: String = tagInfo.tagValueType match {
      case ConstCode.TAG_VALUE_TYPE_STRING => "STRING"
      case ConstCode.TAG_VALUE_TYPE_LONG => "BIGINT"
      case ConstCode.TAG_VALUE_TYPE_DECIMAL => "DECIMAL(16,2)"
      case ConstCode.TAG_VALUE_TYPE_DATE => "STRING"
    }

    val properties: Properties = MyPropertiesUtil.load("config.properties")
    val userProfileDbName = properties.getProperty("user-profile.dbname")
    val wareHouseDbName = properties.getProperty("data-warehouse.dbname")
    val hdfsStorePath = properties.getProperty("hdfs-store.path")

    //  create table  $tableName
    // 字段： ( uid string, tag_value $tagValueType     )
    //  partitioned by (dt string)
    //  ROW FORMAT DELIMITED FIELDS TERMINATED BY '\\t'
    // LOCATION  $hdfsStorePath/$userProfileDbName/$tableName

    val createTableSQL =
      s"""
         |   create table if  not exists  $userProfileDbName.$tableName
         |    ( uid string, tag_value $tagValueType     )
         |    partitioned by (dt string)
         |   ROW FORMAT DELIMITED FIELDS TERMINATED BY '\\t'
         |   LOCATION  '$hdfsStorePath/$userProfileDbName/$tableName'
     """.stripMargin

    println(createTableSQL)
    sparkSession.sql(createTableSQL)

    //3  组合select SQL 语句
    //  select uid ,  case  query_value  when 'F' then '女' ...  end
    //from  ($TAG_SQL)

    //根据四级标签的映射关系动态组合case when 子句
    // #作业：兼容 带四级标签和 不带四级标签的情况
    var tagValueSQL = ""
   if (taskTagRuleList.size>0){
    val whenThenList: List[String] = taskTagRuleList.map(taskTagRule => s" when  '${taskTagRule.queryValue}' then  '${taskTagRule.subTagValue}' ")
    val whenThenSQL: String = whenThenList.mkString(" ")
     tagValueSQL = s"case query_value $whenThenSQL end "

   }else{
     tagValueSQL =" query_value "
   }
    //#作业 替换sql中的日期标识
    val tagsql: String = taskInfo.taskSql.replace("$dt",taskDate)
    val selectSQL = s"  select uid ,  $tagValueSQL as tag_value from (${tagsql}) tagsql"
    println(selectSQL)

    //4 组合插入sql
    sparkSession.sql(s"use $wareHouseDbName")
    val insertSQL = s"  insert overwrite table $userProfileDbName.$tableName partition (dt='$taskDate') $selectSQL"
    println(insertSQL)
    sparkSession.sql(insertSQL)
  }

}

File: task-sql/src/main/java/com/zxkw/userprofile/app/TaskSQL.java
package com.zxkw.userprofile.app;

import com.zxkw.userprofile.common.bean.TagInfo;
import com.zxkw.userprofile.common.bean.TaskInfo;
import com.zxkw.userprofile.common.bean.TaskTagRule;
import com.zxkw.userprofile.common.constants.ConstCode;
import com.zxkw.userprofile.common.dao.TagInfoDao;
import com.zxkw.userprofile.common.dao.TaskInfoDao;
import com.zxkw.userprofile.common.dao.TaskTagRuleDao;
import com.zxkw.userprofile.common.util.MyPropsUtil;
import org.apache.spark.SparkConf;
import org.apache.spark.sql.SparkSession;

import java.util.List;

/**
 * 开发任务：根据SQL定义计算标签
 * 1. 接收spark-submit中传递的业务参数， 也就是task_id
 * 2. 根据task_id 查询mysql, 获得对应的task_info 、tag_info 、 task_tag_rule信息
 *    也就是标签名称、标签编码、定义的sql 、 获得计算值与标签值的映射关系
 * 3. 生成标签前, 先要检查是否有该标签的表， 如果没有， 要为给标签建立hive数据表
 * 4. 根据sql 和映射 , 动态拼接一个 insert select SQL
 * 5. 根据执行该sql
 */
public class TaskSQL {
    public static void main(String[] args) {
        //0. 执行环境
        SparkConf sparkConf = new SparkConf().setAppName("task_sql_app")
                //仅适用于idea本地;测试或生产集群运行需注释
                .setMaster("local[*]");
            //创建hive支持
        SparkSession sparkSession = SparkSession.builder().config(sparkConf).enableHiveSupport().getOrCreate();

        //1. 接收spark-submit中传递的业务参数， 也就是task_id
        //spark-submit -xx -xx --class xxx  xxx.jar  101  2020-06-14
        //约定第一个参数放task_id , 第二参数方business_date
        String taskId =  args[0];
        String businessDate = args[1];

        //2. 根据task_id查询mysql , 获得对应的task_info , tag_info , task_tag_rule信息
        TagInfo tagInfo = TagInfoDao.getTagInfoByTaskId(taskId);
        System.out.println(tagInfo);
        TaskInfo taskInfo = TaskInfoDao.getTaskInfoById(taskId);
        System.out.println(taskInfo);
        List<TaskTagRule> taskTagRuleList = TaskTagRuleDao.getTaskTagRuleList(taskId);
        System.out.println(taskTagRuleList);

        String tagCode = tagInfo.getTagCode();
        String tagName = tagInfo.getTagName();
        String taskSql = taskInfo.getTaskSql();

        //TODO 补充: 替换SQL中$dt
        taskSql = taskSql.replace("$dt", businessDate);

        //3. 生成标签前,先要检查是否有该标签的表， 如果没有, 要为该标签建立数据表
        //   每张表负责一个标签, 两个字段:  谁 , 值是什么 , 比如性别标签表： uid ,tag_value  , 分区字段 busiDate
        // create table tableName[tagCode]
        // ( uid string , tag_value tagValueType?)
        // partitioned by (dt string)
        // ROW FORMAT DELIMITED FIELDS TERMINATED BY '\\t'
        // location 'hdfs标签主目录/画像库名/表名'

        // 表名 和 表描述
        String tableName = tagCode.toLowerCase();
        String comment = tagName ;

        String fieldType = null ;
        //确定字段类型
        switch (tagInfo.getTagValueType()){
            case ConstCode.TAG_VALUE_TYPE_LONG :
                fieldType = "bigint";
                break ;
            case ConstCode.TAG_VALUE_TYPE_DECIMAL:
                fieldType = "decimal(16,2)";
                break ;
            case ConstCode.TAG_VALUE_TYPE_STRING :
                fieldType = "string";
                break ;
            case ConstCode.TAG_VALUE_TYPE_DATE :
                fieldType = "string";
                break ;
        }

        String hdfsPath = MyPropsUtil.get(ConstCode.HDFS_STORE_PATH);
        String dwDbName = MyPropsUtil.get(ConstCode.DATA_WAREHOUSE_DBNAME);
        String upDbName = MyPropsUtil.get(ConstCode.USER_PROFILE_DBNAME);

        String createTableSql =
                " create table if not exists " + upDbName + "." + tableName +
                        " (" +
                        " uid string , " +
                        " tag_value " + fieldType +
                        " )" +
                        " partitioned by (dt string) " +
                        " row format delimited fields terminated by '\\t'" +
                        " location '" + hdfsPath+ "/" + upDbName + "/" + tableName + "'";

        System.out.println("createTableSQL:" + createTableSql);
        sparkSession.sql(createTableSql);

        //4. 要根据sql和映射， 动态拼接一个insert select sql
        // select uid , case query_value when 'F' then '女' when 'M' then '男' when 'U' then '未知'
        // from
        // (
        // select id as uid , if(gender <> '' , gender, 'U') as query_value from dim_user_zip where dt = '9999-12-31'
        // )

        // streamAPI
        //List<String> whenThenList = taskTagRuleList.stream().map(
        //        taskTagRule -> "when '" + taskTagRule.getQueryValue() + "' then '" + taskTagRule.getSubTagValue() + "'"
        //).collect(Collectors.toList());

        //普通代码

        //TODO  判断是否有四级标签
        String queryValueSql = "";
        if(taskTagRuleList.size() > 0 ){
            //TODO  如果有四级标签，完成标签值到映射转换
            //拼接case when then end
            String caseWhenThenString  = "case query_value " ;
            for (TaskTagRule taskTagRule : taskTagRuleList) {
                caseWhenThenString += (" when '" +taskTagRule.getQueryValue() + "' then '" + taskTagRule.getSubTagValue() + "'");
            }
            caseWhenThenString += " end as query_value ";
            queryValueSql = caseWhenThenString ;
        }else{
            //TODO  如果没有四级标签, 查询值即为标签值
            queryValueSql = " query_value " ;
        }

        // 拼接SelectSQL
        String selectSQL = " select uid , " + queryValueSql  + " from (" +  taskSql  + ")" ;
        System.out.println("selectSQL:" + selectSQL);

        // 拼接InsertSQL
        String insertSQL = " insert overwrite table " + upDbName + "." + tableName + " partition (dt = '"+ businessDate+"') " + selectSQL;
        System.out.println("insertSQL:" + insertSQL );
        sparkSession.sql("use "+ dwDbName );
        sparkSession.sql(insertSQL);
    }
}
